src/dlls/mscoree/unixinterface.cpp

```diff
@@ -87,6 +87,80 @@ static LPCWSTR* StringArrayToUnicode(int argc, LPCSTR* argv)
     return argvW;
 }
 
+static void ExtractStartupFlagsAndConvertToUnicode(
+    const char** propertyKeys,
+    const char** propertyValues,
+    int* propertyCountRef,
+    STARTUP_FLAGS* startupFlagsRef,
+    LPCWSTR** propertyKeysWRef,
+    LPCWSTR** propertyValuesWRef)
+{
+    int propertyCount = *propertyCountRef;
+
+    LPCWSTR* propertyKeysW = new (nothrow) LPCWSTR[propertyCount];
+    ASSERTE_ALL_BUILDS(propertyKeysW != nullptr);
+
+    LPCWSTR* propertyValuesW = new (nothrow) LPCWSTR[propertyCount];
+    ASSERTE_ALL_BUILDS(propertyValuesW != nullptr);
+
+    // Extract the startup flags from the properties, and convert the remaining properties into unicode
+    STARTUP_FLAGS startupFlags =
+        static_cast<STARTUP_FLAGS>(
+            STARTUP_FLAGS::STARTUP_LOADER_OPTIMIZATION_SINGLE_DOMAIN |
+            STARTUP_FLAGS::STARTUP_SINGLE_APPDOMAIN |
+            STARTUP_FLAGS::STARTUP_CONCURRENT_GC);
+    int propertyCountW = 0;
+    for (int propertyIndex = 0; propertyIndex < propertyCount; ++propertyIndex)
+    {
+        auto SetFlagValue = [&](STARTUP_FLAGS flag)
+        {
+            if (strcmp(propertyValues[propertyIndex], "0") == 0)
+            {
+                startupFlags = static_cast<STARTUP_FLAGS>(startupFlags & ~flag);
+            }
+            else if (strcmp(propertyValues[propertyIndex], "1") == 0)
+            {
+                startupFlags = static_cast<STARTUP_FLAGS>(startupFlags | flag);
+            }
+        };
+
+        if (strcmp(propertyKeys[propertyIndex], "CONCURRENT_GC") == 0)
```

Comment:
No problem, thanks @janvorli. Without going into too much detail, I'm quite satisfied with the readability of the code. I see that making a reusable function to extract boolean flags may be useful but to preserve the logic here, something like TryExtractBooleanFlag would have to return true when the property key matches the specified flag, regardless of the property value, and that I think makes it less readable. Or in the form you suggest, the function would be very specific to this particular usage, in which case I feel a lambda is far more readable, as the small amount of logic involved is more local to the uses and greatly simplifies the calls. I'll leave it as is for now, and will reconsider if there is more demand for reusability here, which seems unlikely considering the simplicity.